drivers/modem/gsm_ppp.c

```diff
@@ -1181,26 +1220,231 @@ void gsm_ppp_stop(const struct device *dev)
+				    &response_cmds[0], ARRAY_SIZE(response_cmds), cmd_buffer,
+				    &gsm->sem_response, GSM_CMD_SETUP_TIMEOUT);
+
+	if (ret < 0) {
+		LOG_DBG("Could not set ring indicator setting for sms");
+	}
+
+	if (other == ALWAYS) {
+		LOG_ERR("Incorrect setting for other ring indicator, setting to OFF");
+		other = OFF;
+	}
+
+	gsm_ppp_generate_ring_indicator_at_cmd(cmd_buffer, sizeof(cmd_buffer),
+					       "urc/ri/other", other, other_pulse_duration);
+	ret = modem_cmd_send_nolock(&gsm->context.iface, &gsm->context.cmd_handler,
+				    &response_cmds[0], ARRAY_SIZE(response_cmds), cmd_buffer,
+				    &gsm->sem_response, GSM_CMD_SETUP_TIMEOUT);
+
+	if (ret < 0) {
+		LOG_DBG("Could not set ring indicator setting");
+	}
+}
+
+static const struct setup_cmd sms_configure_cmds[] = {
+	/* Set SMS message format to text */
+	SETUP_CMD_NOHANDLE("AT+CMGF=1"),
+	/* Set SMS message reporting to message the receiver */
+	SETUP_CMD_NOHANDLE("AT+CNMI=2,1"),
+	/* Set SMS message character set */
+	SETUP_CMD_NOHANDLE("AT+CSCS=\"IRA\""),
+};
+
+void gsm_ppp_configure_sms_reception(const struct device *dev)
+{
+	struct gsm_modem *gsm = dev->data;
+	int ret;
+
+	ret = modem_cmd_handler_setup_cmds_nolock(&gsm->context.iface,
+						  &gsm->context.cmd_handler,
+						  sms_configure_cmds,
+						  ARRAY_SIZE(sms_configure_cmds),
+						  &gsm->sem_response,
+						  GSM_CMD_SETUP_TIMEOUT);
+
+	if (ret < 0) {
+		LOG_DBG("Could not set SMS configuration");
+	}
+
+	gsm_ppp_clear_ring_indicator(dev);
+}
+
+static struct gsm_ppp_sms_message sms_message;
+
+/* Handler: <SMS message> */
+MODEM_CMD_DEFINE(on_cmd_atcmd_read_sms)
+{
+	if (strncmp(argv[0], "+CMTI: ", strlen("+CMTI: ")) == 0) {
+		LOG_DBG("SMS message received by modem");
+	} else if (argc >= 1 && strncmp(argv[0], "+CMGL: ", strlen("+CMGL: ")) == 0) {
+		sms_message.index = atoi(strchr(argv[0], ' '));
+
+		unquoted_strncpy(sms_message.status, argv[1], GSM_PPP_SMS_STATUS_LENGTH);
+
+		unquoted_strncpy(sms_message.origin_address, argv[2], GSM_PPP_SMS_OA_LENGTH);
+
+		unquoted_strncpy(sms_message.origin_name, argv[3], GSM_PPP_SMS_OA_NAME_LENGTH);
+
+		unquoted_strncpy(sms_message.date, argv[4], GSM_PPP_SMS_DATE_LENGTH);
+
+		unquoted_strncpy(sms_message.time, argv[5], GSM_PPP_SMS_TIME_LENGTH);
+	} else if(argc == 1) {
+		size_t out_len;
+
+		sms_message.has_data = true;
+		out_len = net_buf_linearize(sms_message.data,
+						sizeof(sms_message.data) - 1,
+						data->rx_buf, 0, len);
+		sms_message.data[out_len] = '\0';
+	} else {
+		LOG_DBG("Invalid SMS received");
+		sms_message.valid = false;
+	}
+	return 0;
+}
+
+static const struct setup_cmd read_sms_cmds[] = {
+	SETUP_CMD_ARGS_MAX("AT+CMGL=\"REC UNREAD\"", "", on_cmd_atcmd_read_sms, 0U, 8U, ","),
+};
+
+struct gsm_ppp_sms_message * gsm_ppp_read_sms(const struct device *dev)
+{
+	struct gsm_modem *gsm = dev->data;
+	int ret;
+
+	memset(&sms_message, 0, sizeof(sms_message));
+	sms_message.valid = true;
```

Comment:
Are you sure it is always valid here? I can image in that if reading out the message fails, that is it not valid. as well. So consider clearing the flag on fail or, only set it at success.

Reply:
Done